File: OS Lab/Assignment 2/c++/1_a.cpp
#include <iostream>
#include <cstdlib>
#include <ctime>
#include <sys/types.h>
#include <unistd.h>

using namespace std;

// Function for child process X
void process_X(int iterations) {
    srand(time(nullptr)); // Seed the random number generator
    for (int i = 0; i < iterations; ++i) {
        double sleep_duration = (rand() % 5 + 1) / 10.0; // Random sleep between 0.1 and 0.5 seconds
        cout << "Process X - Iteration " << i + 1 << ", Sleeping for " << sleep_duration << " seconds" << endl;
        sleep(sleep_duration); // Sleep in second
    }
}

// Function for child process Y
void process_Y(int iterations) {
    srand(time(nullptr)); // Seed the random number generator
    for (int i = 0; i < iterations; ++i) {
        double sleep_duration = (rand() % 5 + 5) / 10.0; // Random sleep between 0.5 and 1.0 seconds
        cout << "Process Y - Iteration " << i + 1 << ", Sleeping for " << sleep_duration << " seconds" << endl;
        sleep(sleep_duration); // Sleep in second
    }
}

int main() {
    int iterations = 10;

    // Create child processes X and Y
    pid_t child_x, child_y;

    child_x = fork(); // Create child process X

    if (child_x == 0) { // In the child process X
        process_X(iterations);
        exit(0);
    } else if (child_x > 0) { // In the parent process
        child_y = fork(); // Create child process Y

        if (child_y == 0) { // In the child process Y
            process_Y(iterations);
            exit(0);
        } else if (child_y > 0) { // In the parent process
            // Wait for both child processes to finish
            int status;
            waitpid(child_x, &status, 0);
            waitpid(child_y, &status, 0);
            cout << "Both processes have finished." << endl;
        } else {
            cerr << "Error creating child process Y." << endl;
            exit(1);
        }
    } else {
        cerr << "Error creating child process X." << endl;
        exit(1);
    }

    return 0;
}

File: OS Lab/Assignment 2/python/1_a.py
import multiprocessing
import random
import time

# Function for child process X
def process_X(iterations):
    for i in range(iterations):
        sleep_duration = random.uniform(0.1, 0.5)  # Random sleep between 0.1 and 0.5 seconds
        print(f"Process X - Iteration {i+1}, Sleeping for {sleep_duration:.2f} seconds")
        time.sleep(sleep_duration)

# Function for child process Y
def process_Y(iterations):
    for i in range(iterations):
        sleep_duration = random.uniform(0.5, 1.0)  # Random sleep between 0.5 and 1.0 seconds
        print(f"Process Y - Iteration {i+1}, Sleeping for {sleep_duration:.2f} seconds")
        time.sleep(sleep_duration)

if __name__ == '__main__':
    iterations = 10

    # Create child processes X and Y
    process_x = multiprocessing.Process(target=process_X, args=(iterations,))
    process_y = multiprocessing.Process(target=process_Y, args=(iterations,))

    # Start the child processes
    process_x.start()
    process_y.start()

    # Wait for both processes to finish
    process_x.join()
    process_y.join()

    print("Both processes have finished.")
